sway-core/src/asm_generation/programs.rs

```diff
@@ -44,26 +45,38 @@ pub(super) struct AbstractEntry {
     pub(super) label: Label,
     pub(super) ops: AbstractInstructionSet,
     pub(super) name: FnName,
-    pub(super) test_decl_ref: Option<DeclRef>,
+    pub(super) test_decl_ref: Option<DeclRef<DeclId<TyFunctionDeclaration>>>,
 }
 
 /// An AllocatedProgram represents code which has allocated registers but still has abstract
 /// control flow.
+#[allow(clippy::type_complexity)]
 pub(super) struct AllocatedProgram {
     kind: ProgramKind,
     data_section: DataSection,
     prologue: AllocatedAbstractInstructionSet,
     functions: Vec<AllocatedAbstractInstructionSet>,
-    entries: Vec<(SelectorOpt, Label, FnName, Option<DeclRef>)>,
+    entries: Vec<(
+        SelectorOpt,
+        Label,
+        FnName,
+        Option<DeclRef<DeclId<TyFunctionDeclaration>>>,
```

Comment:
Some of these types are pretty complex. What are your thoughts on making some type aliases?

Reply:
I'm not sure what would be ergonomic here.

The actual declaration type should surely be included, and I do want some differentation between `DeclRef`s of raw ids versus mixed enums of ids.

Maybe something like ```rust
type DeclRefFunction = DeclRef<DeclId<TyFunctionDeclaration>>;
type DeclRefMixFunctional = DeclRef<FunctionalDeclId>;
```

Reply:
Nice, I personally prefer type aliases so it would get my vote. I'm sure others have opinions about this though.